docs/DafnyRef/UserGuide.md

```diff
@@ -365,7 +365,18 @@ The command emits exit codes of
 
 #### 25.5.1.9. `dafny format` {#sec-dafny-format}
 
-This command is not yet released, but will be a command that formats source code to a consistent style.
+Dafny supports a formatter, which for now only changes the indentation of every line in a Dafny file, so that it conforms
+to the idiomatic Dafny code formatting style.
+For the formatter to work, the file should be parsed correctly by Dafny.
+
+There are four ways to use the formatter:
+
+* `dafny format file1.dfy file2.dfy ...` formats each file provided in arguments.
+* `dafny format --print:- file1.dfy file2.dfy ...` format each file but instead of writing files, output their formatted content in stdout
+* `dafny format` Look up for all the Dafny files recursively in the current repository and format them.
+* `dafny format --check` with the same options as above. It will print a message
+  and a non-zero return code if at least one file is not the same as
+  its formatted version through `dafny format`
```

Comment:
1. I changed the "--print:-" option to a boolean "--print" which was very easy thanks to the new command line
2. dotnet format already does that, and our formatter is pretty stable. I would like to get more opinions on that @robin-aws @keyboardDrummer and @alex-chew  ? Vote 🅰️ if you want to enforce "." and 🅱️  if you think `dafny format` is enough
3. --check does not modify any file, and tells which files are not correctly formatted, like this.
```
$ ./Binaries/Dafny.exe format --check Source/DafnyCore/AST/Formatting.dfy
The file Formatting.dfy needs to be formatted
Error: 1 file need formatting.

$ ./Binaries/Dafny.exe format --check Source/DafnyCore/AST/Formatting.dfy
All files correctly formatted
```
Could we please keep diffs for an upcoming PR based on needs?

Reply:
I think `dafny format` should not try to format everything if no argument is passed, since that seems too implicit of a behavior. Regarding formatting a directory, can't a user use shell wildcards to pass many individual files to the CLI? Why do we need first class directory support?

I do think it should accept individual files or a project file (when it becomes available)

Reply:
To format all dafny files recursively under a directory would take combining with a `find` call. I think it better to have some options to recursively format all dafny files under some directory.

Reply:
You can always use "ls" and "xargs" to provide a specific list of files to format to Dafny.

I'm optimizing for the two most common cases:
* Formatting all the dafny files recursively in a repository as part of a commit process: `dafny format .` (I adopted your suggestion @davidcok  of accepting folders and `dafny format` now returns some usage help)
* Formatting specific dafny files.

Reply:
To do this, there is already the following command:

```
dafny format --stdin --print
```
so I don't think it's worth adding a special case for "-". First and foremost, no one complained that stdin was not supported in the new Dafny version so I'm pretty sure it won't be a pain point for now, and if it is later, someone could always suggest a PR, but I would be skeptical to support ad-hoc use cases.

Reply:
Agree - since --print now takes no argument, we don't need -. But what is the default output  when using --stdin? I think it might as well be stdout (not anything like stdin.dfy). One can always redirect it where one wants.

Reply:
Currently, the use of --stdin requires the use of either --check or --print, otherwise it returns an error and explanation.

Reply:
That's OK, at least until user feedback says differently.